AppCenter/AppCenter/Internals/Util/MSACEncrypter.m

```diff
@@ -60,37 +76,80 @@ - (NSString *_Nullable)decryptString:(NSString *)string {
 
 - (NSData *_Nullable)decryptData:(NSData *)data {
 
-  // Separate cipher prefix from cipher.
-  NSRange dataRange = NSMakeRange(0, [data length]);
-  NSData *separatorAsData = [kMSACEncryptionMetadataSeparator dataUsingEncoding:NSUTF8StringEncoding];
-  size_t metadataLocation = [data rangeOfData:separatorAsData options:0 range:dataRange].location;
-  NSString *metadata;
-  if (metadataLocation != NSNotFound) {
-    NSData *subdata = [data subdataWithRange:NSMakeRange(0, metadataLocation)];
-    metadata = [[NSString alloc] initWithData:subdata encoding:NSUTF8StringEncoding];
-  }
-  NSData *key;
+  // Get secret key.
+  NSData *secretKey;
+
+  // Load metadata.
+  size_t metadataLocation = [self loadMetadataLocation:data];
+  NSString *metadata = [self loadMetadata:data metadataLocation:metadataLocation];
+  NSString *keyTag = [metadata componentsSeparatedByString:kMSACEncryptionMetadataInternalSeparator][0];
+
+  // Load data.
   NSData *initializationVector;
   NSData *cipherText;
+  NSData *hMac;
   if (metadata) {
-
-    // Extract key from metadata.
-    NSString *keyTag = [metadata componentsSeparatedByString:kMSACEncryptionMetadataInternalSeparator][0];
-
-    // Metadata, separator, and initialization vector.
-    size_t cipherTextPrefixLength = metadataLocation + 1 + kCCBlockSizeAES128;
-    NSRange cipherTextRange = NSMakeRange(cipherTextPrefixLength, [data length] - cipherTextPrefixLength);
     NSRange ivRange = NSMakeRange(metadataLocation + 1, kCCBlockSizeAES128);
-    initializationVector = [data subdataWithRange:ivRange];
-    cipherText = [data subdataWithRange:cipherTextRange];
-    key = [self getKeyWithKeyTag:keyTag];
```

Comment:
Why did you remove this line. `[self getKeyWithKeyTag:keyTag]` is called multiple times below. Let's save it to var.

Reply:
Nope, it is called only once.